### PR TITLE
[TF] Suppress TensorFlow runtime logging unless specified by the user

### PR DESCRIPTION
**Pull request by rxwei**
TensorFlow logging has become noise to the user. We set environment variable "TF_CPP_MIN_LOG_LEVEL" to “3” to suppress such logging without overriding any user-given log level. We do this at TF (or _ExecutionContext) initialization time in the compiler runtime.

The following:
```console
2018-10-02 14:54:52.724042: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.2 AVX AVX2 AVX512F FMA
2018-10-02 14:54:52.739486: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.741925: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.767617: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.769572: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.771369: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.773258: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.775125: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.776862: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.778672: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.780525: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.782194: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.784092: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.785902: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.787656: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
2018-10-02 14:54:52.789460: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:344] Starting optimization for grappler item: tf_graph
[[1.07379355e+24, 1.5649758e+24], [2.3474636e+24, 3.421257e+24]]
```

... becomes 

```console
[[1.07379355e+24, 1.5649758e+24], [2.3474636e+24, 3.421257e+24]]
```

### CODE REVIEW

**Review by lattner — APPROVED** [review]
LGTM, but is there going to be some user-accessible function call to explicitly manage this?  Maybe the logging level should be a global variable set by a "TensorFlow.setLoggingLevel(1)" sort of call, and the 1 then gets passed in here?

**Review by mhong — APPROVED** [review]
I need to investigate more into why this grappler message is printed, given the source code shows it's under VLOG(1) and not LOG(INFO). But pls feel free to land this patch.

**Comment by mhong** [comment]
I built a local toolchain via:
`SWIFT_PACKAGE=tensorflow_linux,no_test swift/utils/build-toolchain local.swift`

I cannot reproduce the grappler string:
```
$ ../build/buildbot_linux/swift-linux-x86_64/bin/swiftc -O test/TensorFlow/tmp.swift && ./tmp
2018-10-03 17:51:42.228577: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-10-03 17:51:42.426340: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:02:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2018-10-03 17:51:42.538552: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 1 with properties: 
name: Quadro K1200 major: 5 minor: 0 memoryClockRate(GHz): 1.0325
pciBusID: 0000:03:00.0
totalMemory: 3.91GiB freeMemory: 1.67GiB
2018-10-03 17:51:42.538597: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1496] Ignoring visible gpu device (device: 1, name: Quadro K1200, pci bus id: 0000:03:00.0, compute capability: 5.0) with Cuda multiprocessor count: 4. The minimum required count is 8. You can adjust this requirement with the env var TF_MIN_GPU_MULTIPROCESSOR_COUNT.
2018-10-03 17:51:42.538607: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2018-10-03 17:51:42.770785: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-10-03 17:51:42.770818: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 1 
2018-10-03 17:51:42.770839: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N N 
2018-10-03 17:51:42.770843: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 1:   N N 
2018-10-03 17:51:42.771037: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7526 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:02:00.0, compute capability: 6.1)
2018-10-03 17:51:42.781563: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1496] Ignoring visible gpu device (device: 1, name: Quadro K1200, pci bus id: 0000:03:00.0, compute capability: 5.0) with Cuda multiprocessor count: 4. The minimum required count is 8. You can adjust this requirement with the env var TF_MIN_GPU_MULTIPROCESSOR_COUNT.
2018-10-03 17:51:42.781588: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2018-10-03 17:51:42.781615: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-10-03 17:51:42.781622: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 1 
2018-10-03 17:51:42.781628: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N N 
2018-10-03 17:51:42.781633: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 1:   N N 
2018-10-03 17:51:42.781804: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7526 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:02:00.0, compute capability: 6.1)
```

The messages from `gpu_device.cc` are via LOG(INFO).

@marcrasi , Marc, how did you build the Linux toolchain?

**Comment by mhong** [comment]
`tmp.swift` has a trivial program like:
```swift
import TensorFlow
public func loop() {
  var i: Int32 = 0
  while (i < 3) {
    _ = Tensor(i)
    i += 1
  }
}
loop()
```

File: stdlib/public/TensorFlow/CompilerRuntime.swift [hunk 1]
```diff
@@ -264,6 +264,9 @@ public final class _ExecutionContext {
   init() {
     configureRuntimeFromEnvironment()
 
+    // Suppress TensorFlow logging, unless the user specified a log level.
+    setenv("TF_CPP_MIN_LOG_LEVEL", "3", /*override*/ 0)
```
[the remark this answers is not in the document]
Comment (rxwei) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: That's fine with me. FWIW, it's inconsistent with the name of that function. It's not configuring the runtime from environment. Quite the opposite, it's configuring the environment.

[the remark this answers is not in the document]
Comment (rxwei) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: Moreover, `configureRuntimeFromEnvironment()` is being called the second time in `_TensorComputation.init`. I'd prefer to leave it as is until there are other env-configuring operations.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
> is there going to be some user-accessible function call to explicitly manage this? Maybe the logging level should be a global variable set by a "TensorFlow.setLoggingLevel(1)" sort of call, and the 1 then gets passed in here?

For now, the user can set the environment variable directly. Adding a new API to set it is functionally separate from this patch and will be considered later.

File: stdlib/public/TensorFlow/CompilerRuntime.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhong) [reply stdlib/public/TensorFlow/CompilerRuntime.swift]: SG

**Pull request merged**